### PR TITLE
WFLY-6402 EJBs accessible too early (spec violation)

### PR DESCRIPTION
**Pull request by kurobako**
https://issues.jboss.org/browse/WFLY-6402
EJBs accessible too early (spec violation)

now all startup beans in deployment should finish their postconstruct methods before external calls would be allowed to proceed

### CODE REVIEW

**Comment by bmaxwell** [comment]
retest this please!!

**Comment by dmlloyd** [comment]
retest this please

**Comment by dmlloyd** [comment]
Seems like this PR is failing consistently.

**Comment by kurobako** [comment]
@dmlloyd both failures occur while running SendMessagesTestCase times out:

[15:06:16][org.wildfly:wildfly-ts-integ-basic] Running org.jboss.as.test.integration.ejb.mdb.containerstart.SendMessagesTestCase
[16:36:36][org.wildfly:wildfly-ts-integ-basic] The build Pull Request::Linux #9792 {buildId=96690} has been running for more than 100 minutes. Terminating...

I wonder why test with security manager succeeds, though. Afaik the spec forces us to hold MDB calls until @Startup @PostConstruct is done as well, right? Could it be I miss some aspect of how MDB work in such case?

**Comment by dmlloyd** [comment]
I'm not sure, it's pretty strange.  Perhaps there's some race condition, and using a security manager upsets the timing to just the perfect degree that the race condition does not occur.  This also might explain why the PR passed on the EAP branch.

File: ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/merging/StartupMergingProcessor.java [hunk 1]
```diff
@@ -51,6 +51,7 @@ protected void handleAnnotations(final DeploymentUnit deploymentUnit, final EEAp
             if (data != null) {
                 if (!data.getClassLevelAnnotations().isEmpty()) {
                     description.initOnStartup();
+                    description.getModuleDescription().registerStartupBean();
```
Comment (stuartwdouglas) [inline ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/merging/StartupMergingProcessor.java]: Why does this only handle annotations, and not the deployment descriptor as well?

File: ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/StartupAwaitDeploymentUnitProcessor.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Adds  StartupAwaitInterceptor to exposed methods of EJB, forcing users to wait until all startup beans in the deployment are done with post-construct methods.
+ * @author Fedor Gavrilov
+ */
+// adding an abstraction for the whole deployment unit to depend on while blocking external client calls is a better solution probably
+// it requires a lot of rewriting in EJB code right now, hence this class to satisfy EJB 3.1 spec, section 4.8.1
+// feel free to remove this class as well as StartupAwaitInterceptor and StartupCountDownInterceptor when if easier way to satisfy spec will appear
+public class StartupAwaitDeploymentUnitProcessor implements DeploymentUnitProcessor {
+    private static final Set<MethodIntf> INTFS = EnumSet.of(MethodIntf.MESSAGE_ENDPOINT, MethodIntf.REMOTE, MethodIntf.SERVICE_ENDPOINT);
```
Comment (stuartwdouglas) [inline ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/StartupAwaitDeploymentUnitProcessor.java]: This is not enought to satisfy the spec. Local invocation could be delivered through a servlet that injects the local interface for example.

File: ejb3/src/main/java/org/jboss/as/ejb3/component/singleton/StartupCountDownInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+public class StartupCountDownInterceptor implements Interceptor {
+    private final CountDownLatch latch;
+
+    StartupCountDownInterceptor(final CountDownLatch countDownLatch) {
+        this.latch = countDownLatch;
+    }
+
+    @Override
+    public Object processInvocation(final InterceptorContext context) throws Exception {
+        latch.countDown();
```
Comment (stuartwdouglas) [inline ejb3/src/main/java/org/jboss/as/ejb3/component/singleton/StartupCountDownInterceptor.java]: This counts down before the bean is actually initialised

**Comment by stuartwdouglas** [comment]
There are a number of fairly fundamental problems here:

1) It does not actually fulfill the spec, as external requests can still be delivered though other means, e.g. a web request into a Servlet that injects a local interface
2) If 1) was fixed then any use of another EJB by a Servlet in @Startup would immediately result in a deadlock, this would break many existing applications

I think this could be worked around by introducing a ThreadLocal that tracks if an invocation is a startup post construct invocation. 

Basically StartupCountDownInterceptor would set this to true, then clear it after the request. If this ThreadLocal was set then the other interceptor would not wait on the latch but just let the request through. 

Either way I still don't really like this, it would be better to stop all requests at the relevant endpoints until deployment is complete, much like we already do with graceful shutdown.

**Comment by kurobako** [comment]
retest this please

**Comment by kurobako** [comment]
Build log for failed build contains thread dump showing this:

java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0xcb512288> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at org.jboss.as.ejb3.deployment.processors.StartupAwaitInterceptor.processInvocation(StartupAwaitInterceptor.java:22)

(a direct call to CountDownLatch from StartupAwaitInterceptor) which is just not present in the current version of PR. Currently StartupCountdown class is used as an intermediate between these two. Probably CI is having trouble getting latest changes from amended commits or something.

**Comment by kurobako** [comment]
retest this please

**Comment by n1hility** [comment]
I agree that fundamental problem is that we should suspend at the remote endpoint level (e.g. no servlet, no remoting, etc)

**Comment by kurobako** [comment]
@n1hility doing that would be a proper solution for the problem, and that is what people are going to do for wildfly eventually. Current PR is just a way to fix spec violation until the code handling graceful startup arrives
